plugin/src/main/java/io/deephaven/plugin/type/ObjectType.java

```diff
@@ -46,7 +47,85 @@ public interface ObjectType extends Plugin {
      * @param out the output stream
      * @throws IOException if an IO exception occurs
      */
-    void writeTo(Exporter exporter, Object object, OutputStream out) throws IOException;
+    default void writeTo(Exporter exporter, Object object, OutputStream out) throws IOException {
+        if (supportsBidiMessaging(object) == Kind.BIDIRECTIONAL) {
+            // internal error, shouldn't have called this
+            throw new IllegalStateException(
+                    "Do not call writeTo if supportsBidiMessaging returns true, but writeTo is not implemented");
+        } else {
+            // incorrect implementation
+            throw new IllegalStateException("ObjectType implementation returned false for supportsBidiMessaging");
+        }
+    }
+
+    /**
+     * A stream of messages, either sent from the server to the client, or client to the server. ObjectType plugin
+     * implementations can provide an implementation of this interface for each incoming stream to invoke as messages
+     * arrive, and will likewise be given an instance of this interface to be able to send messages to the client.
+     */
+    interface MessageStream extends AutoCloseable {
+        /**
+         * Transmits data to the remote end of the stream. This can consist of a binary payload and references to
+         * objects on the server.
+         *
+         * @param payload the binary data sent to the remote implementation
+         * @param references server-side object references sent to the remote implementation
+         */
+        void onMessage(ByteBuffer payload, Object[] references);
+
+        /**
+         * Closes the stream on both ends. No further messages can be sent or received.
+         */
+        void close();
+    }
+
+    /**
+     * Signals creation of a client stream to the provided object. The returned MessageStream implementation will be
+     * called with each received message from the server, and can call the provided connection instance to send messages
+     * as needed to the client.
+     * 
+     * @param object the object to create a connection for
+     * @param connection a stream to send objects to the client
+     * @return a stream to receive objects from the client
```

Comment:
Document the usual requirements for a streaming plugin. The returned MessageStream likely needs to keep track of object + connection (for plugins that react to requests). More sophisticated plugins might also need to associate connection with an internal listener or subscription to object.

Reply:
Aren't your definitions for param and return backwards? Shouldn't `{@data connection}` but used for _receiving_ and the return value for _sending_?